### PR TITLE
Separate jobstate from taskstate and datasetstate

### PR DESCRIPTION
**Pull request by zliu41**
This is a solution to #881.
Summary:
- Each `WorkUnitState`/`TaskState` now contains a `JobState` reference, rather than a copy of all properties in `JobState`. `WorkUnitState.getProp()` will first look for the property in its own properties, then in its `WorkUnit`, then in its `JobState`.
- `DatasetState` no longer contains properties in `JobState`, and no longer supports `setProp` or `getProp` methods (will throw `UnsupportedOperationException`). A `DatasetState` can only contain a few specific properties such as `dataset.urn`, and they are accessed via specific setter and getter methods (e.g., `setDatasetUrn()` and `getDatasetUrn()`).
 - The rationale is the following: a `DatasetState` is created in the driver after the job is done. At this point there shouldn’t be a need to look for, change or add any general properties, especially since the driver is not something pluggable like the constructs, where different people can plug in an arbitrarily customized one for their use cases. Also, the driver already has a copy of `JobState`. If it wants to query any properties it can directly query the `JobState`.
 - This means that properties in the `JobState`, which are properties from the job config file and the global property file, will not be persisted into the state store. Since these are static properties rather than runtime properties (runtime properties are in `TaskState`s), there's no need to persist them.

These changes result in a ~90% reduction in memory consumption of a Gobblin job.

### CODE REVIEW

**Comment by zliu41** [comment]
@chavdar @sahilTakiar @ydai1124 @ibuenros can you review?

**Comment by chavdar** [comment]
LGTM

File: gobblin-api/src/main/java/gobblin/configuration/WorkUnitState.java [hunk 1]
```diff
@@ -287,21 +311,22 @@ public boolean equals(Object object) {
     }
 
     WorkUnitState other = (WorkUnitState) object;
-    return ((this.workunit == null && other.workunit == null)
-        || (this.workunit != null && this.workunit.equals(other.workunit))) && super.equals(other);
+    return ((this.workUnit == null && other.workUnit == null)
```
Comment (sahilTakiar) [inline gobblin-api/src/main/java/gobblin/configuration/WorkUnitState.java]: Should this method and `hashCode()` be updated to include the `jobState`. Alternatively, we could just use lombok's `@EqualsAndHashCode`

File: gobblin-runtime/src/main/java/gobblin/runtime/JobState.java [hunk 2]
```diff
@@ -564,8 +564,12 @@ public DatasetState newDatasetState(boolean fullCopy) {
...
-   * A subclass of {@link JobState} that is used to represent dataset states. This class is currently
-   * identical to {@link JobState} except that the name is more meaningful and less confusing.
+   * A subclass of {@link JobState} that is used to represent dataset states.
+   *
+   * <p>
+   *   A {@code DatasetState} does <em>not</em> contain any properties. Operations such as {@link #getProp(String)}
+   *   and {@link #setProp(String, Object)} are not supported.
+   * </p>
    */
   public static class DatasetState extends JobState {
```
Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/JobState.java]: Does `DatasetState` need to extend `JobState` anymore? I'm not sure why it does in the first place, a `DatasetState` should need to know things like `jobId` and doesn't really have a `startTime` or an `endTime`.

Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/JobState.java]: In fact, given than the `getProp` and `setProp` methods throw `UnsupportedOperationException` does this even need to extend `State`?

File: gobblin-runtime/src/main/java/gobblin/runtime/TaskState.java [hunk 3]
```diff
@@ -83,21 +83,20 @@
...
-  public TaskState() {
-  }
+  public TaskState() {}
 
   public TaskState(WorkUnitState workUnitState) {
     // Since getWorkunit() returns an immutable WorkUnit object,
     // the WorkUnit object in this object is also immutable.
-    super(workUnitState.getWorkunit());
+    super(workUnitState.getWorkunit(), workUnitState.getJobState());
     addAll(workUnitState);
```
Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/TaskState.java]: We should also fix that so that a `TaskState` only has a pointer to its `WorkUnitState`, rather than making a full copy of all the properties in the `WorkUnitState`.

On an unrelated note, whats the difference between `TaskState` and `WorkUnitState`? I don't see any functional reason why the two can't be merged, they both seem to try to track some runtime information. But thats a discussion for another day.

File: gobblin-runtime/src/main/java/gobblin/runtime/JobState.java (hunk 2 above)
Comment (zliu41) [inline gobblin-runtime/src/main/java/gobblin/runtime/JobState.java]: Because dataset states need to be persisted into the state store, which, by definition, only takes `State` objects. Also, extending `State` and `JobState` enables it to inherit methods such as `read`,  `write`, `addTaskState`, etc.

File: gobblin-runtime/src/main/java/gobblin/runtime/TaskState.java (hunk 3 above)
Comment (zliu41) [inline gobblin-runtime/src/main/java/gobblin/runtime/TaskState.java]: I agree, I think `WorkUnitState` and `SourceState` are useless, they are the same as `TaskState` and `JobState`, respectively. Also, `Extract` should also be deprecated as it is conceptually identical as `Dataset`. These changes will be part of another PR.

**Comment by zliu41** [comment]
@sahilTakiar updated 53b908f

File: gobblin-runtime/src/main/java/gobblin/runtime/JobState.java (hunk 2 above)
Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/JobState.java]: Yes it will probably require some more re-factoring, so it can be done later.

**Comment by sahilTakiar** [comment]
LGTM!
